kbs-config.toml: introduce admin_api_ro flag to forbid set_* operations

esposem opened the pull request:
In operator-based deployments, kbs-client (and in general the admin API)  should not be allowed to try and modify the resources, for two reasons:
    
    1. trustee runs in a container so the fs is r/o
    2. management of such resources is by design delegated to kubernetes configmaps

Therefore add admin_api_ro flags defaulting to false under the [admin]  section of kbs-config.toml.

esposem committed:
introduce admin_api_ro flag to restrict operations to only r/o

In operator-based deployments, kbs-client (and in general the admin API)
should not be allowed to try and modify the resources, for two reasons:

1. trustee runs in a container so the fs is r/o
2. management of such resources is by design delegated to kubernetes
   configmaps

Therefore add admin_api_ro flags defaulting to false under the [admin]
section of kbs-config.toml.

Signed-off-by: Emanuele Giuseppe Esposito <eesposit@redhat.com>

mkulke reviewed (commented):
thansk! I think this flag is necessary feature atm. a couple of questions on cosmetics and semantics:

- should we call the toggle and variable/field in the code `admin_api_read_only`? `ro` is not very grepable, and it's unidiomatic in rust to sacrifice characters for clarity.
- Is it only the admin_api or the whole state of the service that is read_only? because only the admin can change state, we can drop the `admin_api_...` prefix.
- it looks like we apply the restriction to POST methods. are we using PUT/DELETE/PATCH to alter the state? maybe we can turn it around and assert GET | HEAD when the read_only toggle is set? (and hope that no one in a plugin corner performs state changes with GET 🤞)
- AFAIU, auth_public_key=Some(...) would be overridden by read_only=true. Can we change the internal representation (not the user toggles) of AdminConfig to reflect that by using an auth enum: 

```rust
enum AdminAuth {
   None,
   PubKey(PathBuf),
   ReadOnly,
}

pub struct AdminConfig {
   pub insecure_api: bool,
   pub auth: AdminAuth,
}
```

mkulke reviewed (approved):
looks good, maybe run the formatter to satisfy the ci

mkulke commented:
> basically the storage way to upload a rsvp via kbs-config.toml bypasses all security checks

I don't think this is an argument about security. it's rather about processing-functionality available in the rvps-tool that is absent when using the trustee operator.

(the overarching question, if I understood it correctly, would be whether it's desirable at all to use the trustee operator for deployment and management. see [here](https://github.com/confidential-containers/trustee/issues/904). my take: afaik the operator is currently the closest to anything resembling a production deployment, since authn/z, state and other non-functional aspects are still WIP in trustee itsself)

esposem committed:
rvps: add flag to restrict operations to be read-only

Introduce rvps read_only flag, that forces the rvps server to only allow
queries and not register new reference values.

This is useful for operator/kubernetes/container deployment where the
rvps is run as standalone and therefore not configured via
kbs-config.toml.

Signed-off-by: Emanuele Giuseppe Esposito <eesposit@redhat.com>

esposem commented:
FWIW I added another flag for the rvps standalone server too, but didn't document it in kbs as it doesn't make sense (rvps running in kbs is only accessible via admin apis as far as I understood). This should cover also the rvps-tool case.

Xynnn007 commented on rvps/src/config.rs:
```diff
@@ -15,6 +15,9 @@ pub struct Config {
     pub extractors: Option<ExtractorsConfig>,
+
+    #[serde(default)]
+    pub rvps_read_only: bool,
```
This is in scope of RVPS now, so we might not need the prfix `rvps_`?

fitzthum commented:
Btw @esposem let me clarify something. Is it all operator-based deployments that don't support writing to the filesystem or is that just on OCP?

lmilleri commented:
> Btw @esposem let me clarify something. Is it all operator-based deployments that don't support writing to the filesystem or is that just on OCP?

@fitzthum I think the rationale of having this PR is to make sure that in operator-based deployments, the trustee instance is enforced to be (almost) stateless. This should help the trustee HA use case (every trustee instance must have the same file system content - policies, ref-values, etc). This works when using k8s configmaps and secrets for modifications.
Btw maybe this is just the first iteration, considering the future evolution of trustee, e.g. fs -> db, etc

fitzthum commented:
Ok, now we have more sophisticated admin roles that allow fine-grained control over which endpoints are enabled/

@esposem please re-open if still needed.
